rfcs/0006-transaction.md

```diff
@@ -0,0 +1,366 @@
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn_opts = TransactionOptions::new().with_isolation_level(IsolationLevel::Serializable); // or IsolationLevel::Snapshot
+let txn = kv_store.transaction_with_opts(txn_opts).await.unwrap();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+
+// one could also rollback the transaction by calling `rollback()` or dropping it when it's out of scope.
+txn.rollback().await;
+```
+
+A list of the transaction API is looked like:
+
+```rust
+// it might not really be a trait, but a struct with the methods:
+trait Transaction {
+    async fn commit(&self) -> Result<()>;
+
+    // the rollback operation should not be async, we need to call it in Drop impl
+    fn rollback(&self) -> Result<()>;
+
+    async fn get(&self, key: &[u8]) -> Result<Option<Bytes>>;
+
+    async fn get_with_options(&self, key: &[u8], options: GetOptions) -> Result<Option<Bytes>>;
+
+    async fn put(&self, key: &[u8], value: &[u8]) -> Result<()>;
+
+    async fn put_with_options(&self, key: &[u8], value: &[u8], options: PutOptions) -> Result<()>;
+
+    async fn delete(&self, key: &[u8]) -> Result<()>;
+
+    async fn scan(&self, range: Range<Bytes>) -> Result<Iterator>;
+}
+```
+
+The list of the transaction API is not final, and it's expected to be adjusted during the implementation. The rule of thumb is that the Transaction API should be as aligned as possible to the non-transactioned API in `DB` struct.
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the simplified data structure of the global `Oracle` and the transaction state, with only the fields that are related to the conflict checking:
+
+```rust
+struct Oracle {
```

Comment:
Where would we track the list of running transactions? Shouldn't that be tracked centrally as well, perhaps in the `Oracle` itself?

Reply:
related: https://github.com/slatedb/slatedb/pull/260#discussion_r1885504611

yes, i think it's placed in the central oracle.